File: Programacao_Funcional/slides/alta_ordem/codes/aprovados.hs
data Student = Student String Int

aprovados [] = []
aprovados (x:xs) | score >= 5  = name : aprovados xs
                 | otherwise   = aprovados xs
                 where (Student name score) = x

main = putStr $ unlines $ aprovados xs where
    xs = [ Student "Ana" 8, Student "Beto" 3, Student "Carlos" 5,
           Student "Daniel" 4, Student "Edgar" 7 ]

File: Programacao_Funcional/alta_ordem/codes/aprovados.cpp
#include <bits/stdc++.h>

using namespace std;

struct Student
{
    string name;
    int score;
};

vector<string> aprovados(const vector<Student>& xs)
{
    vector<string> ys;

    for (auto [name, score] : xs)
        if (score >= 5)
            ys.push_back(name);

    return ys;
}

int main()
{
    vector<Student> xs {
        { "Ana", 8 },
        { "Beto", 3 },
        { "Carlos", 5 },
        { "Daniel", 4 },
        { "Edgar", 7 },
    };

    auto ys = aprovados(xs);

    for (auto y : ys)
        cout << y << '\n';

    return 0;
}
